Source/JavaScriptCore/wasm/WasmSectionParser.cpp

```diff
@@ -840,22 +845,146 @@ auto SectionParser::parseRecursionGroup(uint32_t position, RefPtr<TypeDefinition
         for (uint32_t i = 0; i < typeCount; ++i) {
             RefPtr<TypeDefinition> projection = TypeInformation::typeDefinitionForProjection(recursionGroup->index(), i);
             WASM_PARSER_FAIL_IF(!projection, "can't allocate enough memory for recursion group's ", i, "th projection");
+
+            // Checking subtyping requirements has to be deferred until we construct
+            // projections in case recursive references show up in the type.
+            if (TypeInformation::get(types[i]).is<Subtype>())
+                WASM_PARSER_FAIL_IF(!checkSubtypeValidity(projection->unroll()), "structural type is not a subtype of the specified supertype");
+
             m_info->typeSignatures.uncheckedAppend(projection.releaseNonNull());
         }
     } else {
-        // If a type definition is idempotent for the below replacement, then it is non-recursive and doesn't need a projection.
-        if (*firstSignature == firstSignature->replacePlaceholders(recursionGroup->index()))
+        if (!firstSignature->hasRecursiveReference())
             m_info->typeSignatures.uncheckedAppend(firstSignature.releaseNonNull());
         else {
             RefPtr<TypeDefinition> projection = TypeInformation::typeDefinitionForProjection(recursionGroup->index(), 0);
             WASM_PARSER_FAIL_IF(!projection, "can't allocate enough memory for recursion group's 0th projection");
+            if (firstSignature->is<Subtype>())
+                WASM_PARSER_FAIL_IF(!checkSubtypeValidity(projection->unroll()), "structural type is not a subtype of the specified supertype");
             m_info->typeSignatures.uncheckedAppend(projection.releaseNonNull());
         }
     }
 
     return { };
 }
 
+bool SectionParser::checkStructuralSubtype(const TypeDefinition& subtype, const TypeDefinition& supertype)
+{
+    ASSERT(subtype.is<FunctionSignature>() || subtype.is<StructType>() || subtype.is<ArrayType>());
```

Comment:
In the recursive case, why can't these be projections? Along those lines, this also seems to be wrong? Any one of these struct fields could be a Projection, in which case it would need to be compared syntactically instead of expanded(), right?

Reply:
Yeah good question, so the reason the projection can be ignored here is that there are only two places where this function gets called. In one case, it's in `parseSubtype` and in that case it will only call `checkStructuralSubtype` if the parsing is not happening in a recursion group (this is what the enum class `DeferSubtypeCheck` is for). In that case, `subtype` is always a concrete type and not a projection.

The second case is in `parseRecursionGroup`, which calls `checkSubtypeValidity` on any `sub` declarations that occur in the recursion group. This calls `unroll` explicitly before passing the subtype onto `checkStructuralSubtype`, so the projection is also unrolled away here and won't be seen by `checkStructuralSubtype`.

There's a test that should exercise this case too (a `sub` where the parent index is a recursive type): https://github.com/WebKit/WebKit/blob/ecb7dfd6d0257c271e34c14ff42327c521b284f3/JSTests/wasm/gc/sub.js#L83